opentelemetry-sdk/src/log/log_emitter.rs

```diff
@@ -0,0 +1,236 @@
+use super::{BatchLogProcessor, Config, LogProcessor, LogRecord, LogRuntime, SimpleLogProcessor};
+use crate::{
+    export::log::{LogData, LogExporter},
+    resource::{EnvResourceDetector, SdkProvidedResourceDetector},
+    Resource,
+};
+use opentelemetry_api::{log::LogResult, InstrumentationLibrary};
+use std::{
+    borrow::Cow,
+    sync::{Arc, Weak},
+    time::Duration,
+};
+
+#[derive(Debug, Clone)]
+/// Creator for `LogEmitter` instances.
+pub struct LogEmitterProvider {
+    inner: Arc<LogEmitterProviderInner>,
+}
+
+/// Default log emitter name if empty string is provided.
+const DEFAULT_COMPONENT_NAME: &str = "rust.opentelemetry.io/sdk/logemitter";
+
+impl LogEmitterProvider {
+    /// Build a new log emitter provider.
+    pub(crate) fn new(inner: Arc<LogEmitterProviderInner>) -> Self {
+        LogEmitterProvider { inner }
+    }
+
+    /// Create a new `LogEmitterProvider` builder.
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    /// Create a new `LogEmitter`.
+    pub fn log_emitter(&self, name: impl Into<Cow<'static, str>>) -> LogEmitter {
+        self.versioned_log_emitter(name, Some(env!("CARGO_PKG_VERSION")))
+    }
+
+    /// Create a new version `LogEmitter` instance.
+    pub fn versioned_log_emitter(
+        &self,
+        name: impl Into<Cow<'static, str>>,
+        version: Option<&'static str>,
+    ) -> LogEmitter {
+        let name = name.into();
+
+        let component_name = if name.is_empty() {
+            Cow::Borrowed(DEFAULT_COMPONENT_NAME)
+        } else {
+            name
+        };
+
+        LogEmitter::new(
+            InstrumentationLibrary::new(component_name, version.map(Into::into), None),
+            Arc::downgrade(&self.inner),
+        )
+    }
+
+    /// Config associated with this provider.
+    pub fn config(&self) -> &Config {
+        &self.inner.config
+    }
+
+    /// Log processors associated with this provider.
+    pub fn log_processors(&self) -> &Vec<Box<dyn LogProcessor>> {
+        &self.inner.processors
+    }
+
+    /// Force flush all remaining logs in log processors and return results.
+    pub fn force_flush(&self) -> Vec<LogResult<()>> {
+        self.log_processors()
+            .iter()
+            .map(|processor| processor.force_flush())
+            .collect()
+    }
+
+    /// Shuts down this `LogEmitterProvider`, panicking on failure.
+    pub fn shutdown(&mut self) -> Vec<LogResult<()>> {
```

Comment:
We can implement the `shutdown` on Drop?

Reply:
Good idea, added an initial implementation that tries to gracefully handle errors.